File: JavaScript/queue/queue.js
class QueueNode {
	constructor(data) {
		this.data = data;
		this.next = null;
	}
}

class Queue {
	constructor(elems = null) {
		this._front = null;
		this._back = null;
		this._length = 0;
		if (elems) {
			for (let e in elems) {
				this.enqueue(e);
			}
		}
	}
	
	get length() {
		return this._length;
	}
	
	isEmpty() {
		return this._length == 0;
	}
	
	enqueue(e) {
		if (this._length == 0) {
			this._back = new QueueNode(e);
			this._front = this._back;
		}
		else {
			let temp = new QueueNode(e);
			this._back.next = temp;
			this._back = temp;
		}
		this._length++;
	}
	
	peek() {
		if (this._front == null) {
			return null;
		}
		return this._front.data;
	}
	
	dequeue() {
		let data = this.peek();
		if (this._front == this._back) {
			this.clear();
		}
		else {
			this._front = this._front.next;
			this._length++;
		}
		return data;
	}
	
	clear() {
		this._front = null;
		this._back = null;
		this._length = 0;
	}
}

File: Python/queue/queue.py
class QueueNode:
	def __init__(self, data):
		self.data = data
		self.next = None


class Queue:
	def __init__(self, elems = None):
		self._front = None
		self._back = None
		self._length = 0
		if elems:
			for e in elems:
				self.enqueue(e)
	
	def __len__(self):
		return self._length
	
	def __contains__(self, e):
		return self.contains(e)
	
	def empty(self):
		return self._length == 0
	
	def enqueue(self, e):
		if self._length == 0:
			self._back = QueueNode(e)
			self._front = self._back
		else:
			temp = QueueNode(e)
			self._back.next = temp
			self._back = temp
		self._length += 1
	
	def peek(self):
		return getattr(self._front, 'data', None)
	
	def dequeue(self):
		data = self.peek()
		if self._front == self._back:
			self.clear()
		else:
			self._front = self._front.next
			self._length -= 1
		return data
	
	def contains(self, e):
		current = self._front
		while current:
			if current.data == e:
				return True
			current = current.next
		return False
	
	def clear(self):
		self._front = None
		self._back = None
		self._length = 0
